### PR TITLE
Ease the debugging of internal pickle errors.

### PR DESCRIPTION
**Pull request by Marco-Sulla**
All pickle error messages was a generic "cannot pickle 'type' object". Added <s>some explaining for every individual error.</s> `assert`s to individuate in debug mode the error.

### CODE REVIEW

**Review by iritkatriel — COMMENTED** [review]
I think a change like this should come with changes to tests.

File: Objects/typeobject.c [hunk 1]
```diff
@@ -4344,7 +4344,8 @@ _PyObject_GetState(PyObject *obj, int required)
 
         if (required && Py_TYPE(obj)->tp_itemsize) {
             PyErr_Format(PyExc_TypeError,
-                         "cannot pickle '%.200s' object",
+                         "cannot pickle '%.200s' object: expected " \
+                         "fixed-length instance (tp_itemsize != 0)",
```
Comment (iritkatriel) [inline Objects/typeobject.c]: An alternative message here could be "missing __getstate__ on variable-length object".

**Comment by Marco-Sulla** [comment]
> I think a change like this should come with changes to tests.

@iritkatriel Do you mean I should change everything that comes from `grep -R "cannot pickle" .`?

File: Objects/typeobject.c (hunk 1 above)
Comment (Marco-Sulla) [reply Objects/typeobject.c]: Done, thank you.

**Comment by iritkatriel** [comment]
> > I think a change like this should come with changes to tests.
> 
> @iritkatriel Do you mean I should change everything that comes from `grep -R "cannot pickle" .`?

It's important to make sure that the lines you are changing are even covered by tests. If not, then you should add whatever tests are missing. If they exist, then you can change the expected message so that your change is verified, and the PR reviewer can see that there is test coverage and that the message fits the test case.

You also need to create a ticket on bugs.python.org and link it to this PR.

**Comment by Marco-Sulla** [comment]
> It's important to make sure that the lines you are changing are even covered by tests

Not sure how to reproduce all these exceptions. I probably have to create buggy `CPython` types. Must I create them in `_pickle` and test them in `test_pickle`?

**Comment by iritkatriel** [comment]
Some exceptions never make it to users (and sometimes they are sanity checks that can't even happen). In both cases they are not worth changing. Just change the ones you know how to produce.

**Comment by Marco-Sulla** [comment]
> Just change the ones you know how to produce.

OK, I have only to remember how I got one of these errors...
In the meanwhile, I changed the messages with `assert`s, since correctly Serhiy Storchaka in the bug report [pointed me out](https://bugs.python.org/issue41901?@ok_message=msg%20377778%20created%0Aissue%2041901%20message_count%2C%20messages%20edited%20ok&@template=item#msg377763) that is not correct to expose those details to the end user.

Sorry for the confusion.

**Comment by iritkatriel** [comment]
Serhiy made a good point, but I don't quite see how the asserts are going to help. "Internal debugging" sounds like debugging python itself. But the problem you are trying to solve is that a user type is unpicklable and the user is not getting an indication why.

**Comment by Marco-Sulla** [comment]
> But the problem you are trying to solve is that a user type is unpicklable

Well, no, I reproduced this bug trying to add a builltin, CPython type. The error messages was useful to me to identify the problem. I think it was I wrong the size in the `PyTypeObject` creation.

**Comment by Marco-Sulla** [comment]
It seems it fails the tests... probably for the `assert`s themselves.
Maybe I could add `fprintf`s, but they sounds to me a little tricky.

I think it's better to close it.

**Pull request closed without merging**